File: Python/ReconstructItinerary.py
from typing import List
from collections import defaultdict

class ReconstructItinerary:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        graph = defaultdict(list)
        for ticket in tickets:
            src, dest = ticket
            graph[src].append(dest)
        
        for node in graph:
            graph[node].sort()
        
        result = []
        self.dfs("JFK", result, graph)
        
        return result
    
    def dfs(self, src_node: str, result: List[str], graph: dict):
        if src_node in graph:
            neighbours = graph[src_node]
            while neighbours:
                neigh = neighbours.pop(0)
                self.dfs(neigh, result, graph)
        
        result.insert(0, src_node)

File: Cpp/ReconstructItinerary.cpp
class ReconstructItinerary {
public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        unordered_map<string, vector<string>> graph;
        for(auto ticket : tickets) {
            string src = ticket[0];
            string dest = ticket[1];
            graph[src].push_back(dest);
        }
        
        for(auto& node : graph) {
            sort(node.second.begin(), node.second.end());
        }
        
        vector<string> result;
        dfs("JFK", result, graph);
        
        reverse(result.begin(), result.end());
        return result;
    }
    
    void dfs(string src_node, vector<string>& result, unordered_map<string, vector<string>>& graph) {
        if(graph.find(src_node) != graph.end()) {
            auto neighbours = graph[src_node];
            while(!neighbours.empty()) {
                string neigh = neighbours[0];
                neighbours.erase(neighbours.begin());
                dfs(neigh, result, graph);
            }
        }
        
        result.push_back(src_node);
    }
};
